### PR TITLE
[on hold] [core] Change method name 'size' to 'getNumViolations', save the old method as deprecated and update documentation

### PR DESCRIPTION
**Pull request by pdhung3012**
This is the updated pull request from the suggestion of author from discussion #1957

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/Report.java [hunk 1]
```diff
@@ -507,12 +507,23 @@ public int treeSize() {
         return violationTree.size();
     }
 
+
     /**
      * The number of violations.
      *
      * @return number of violations.
```
Comment (pzygielo) [inline pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: Perhaps [@deprecated](https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@deprecated)?

File: pmd-core/src/main/java/net/sourceforge/pmd/Report.java [hunk 2]
```diff
@@ -507,12 +507,23 @@ public int treeSize() {
...
     public int size() {
+        return getNumViolations();
+    }
+
+    /**
+     * The number of violations.
+     *
+     * @return number of violations.
+     */
+    public int getNumViolations() {
```
Comment (pzygielo) [inline pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: Why not `getNumberOfViolations`? Then the whole javadoc could be dropped.

Comment (pdhung3012) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: Hello. I updated the location of @Deprecate tag and change the method name to getNumberOfViolations and remove javadoc

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/1d84ea9e112542cd8e0e7a34affb8c5e/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/Report.java (hunk 2 above)
Comment (pzygielo) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: >  I updated the location of @deprecate

But did you read the article I linked to? I mean javadoc `@deprecated` tag, especially 
>The @deprecated description in the first sentence should at least tell the user when the API was deprecated and what to use as a replacement.

Comment (pdhung3012) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: Oh I see. I updated the java doc of the @deprecated tag

Comment (pzygielo) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: >  I updated the java doc of the `@deprecated` tag

Cool. But you still missed:

>  An @see tag (for Javadoc 1.1) or {@link} tag (for Javadoc 1.2 or later) should be included that points to the replacement method

Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: @pzygelio Thanks for chiming in about `@deprecated`. However, I disagree with your naming suggestion here. `getNumberOfViolations` is clumsy and less easy to read than `getNumViolations`. There's a reason we don't name methods `getTheNumberOfViolations`: it's noise, ie irrelevant cognitive load. Besides, whatever the name, javadoc should be present.

Comment (pdhung3012) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: > > I updated the java doc of the `@deprecated` tag
> 
> Cool. But you still missed:
> 
> > An @see tag (for Javadoc 1.1) or {@link} tag (for Javadoc 1.2 or later) should be included that points to the replacement method

Thank you. I create the @link tag for the replacement method

Comment (pdhung3012) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: > @pzygelio Thanks for chiming in about `@deprecated`. However, I disagree with your naming suggestion here. `getNumberOfViolations` is clumsy and less easy to read than `getNumViolations`. There's a reason we don't name methods `getTheNumberOfViolations`: it's noise, ie irrelevant cognitive load. Besides, whatever the name, javadoc should be present.

Sothat, should I switch to the older name getNumOfViolations?

Comment (pzygielo) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: > There's a reason we don't name methods getTheNumberOfViolations:

I agree for the _the_ determiner, but I do not about `getNumViolations` being easier to read.  `getNumViolations` (for me) sounds like _get violations of number(s)_.
As this is public class and public method I assume it's intended for some external consumption, so to be probably more often read than written. I which case *I* think, my name is better. But I allow to be wrong on that.

> Besides, whatever the name, javadoc should be present.

Serving what purpose? It's bigger noise than extra _berOf_ in the method. With meaningful name, there'd be no need for that documentation. Even repeating method description in `@return` tag seems pointless to me.
Great example in the same file that it's possible to live without javadoc: https://github.com/pmd/pmd/blob/1a5486dc3084447ca567b041f6ece32ef7b927ae/pmd-core/src/main/java/net/sourceforge/pmd/Report.java#L433-L439

Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: @pzygielo I think `getNumX` is such a well-known convention that there's hardly any ambiguity. Eg in JDK 10, there are 322 method declarations with this pattern, versus 55 for `getNumberOfX`. It's also used more in our own sources (19 vs 6).

> Serving what purpose? It's bigger noise than extra berOf in the method. With meaningful name, there'd be no need for that documentation.

As you said this is an API that is designed for external consumption and we'd like to have javadocs for those to ease their use. Possible uses of a javadoc comment here:
* Clarify the occasional confusion you describe with `get violations of number(s)`, by simply saying this is the "number of violations" (if the `int` return type is not explicit enough)
* Specify that this is the number of elements yielded by the `iterator` method

As you say this code will be read more than written, but I think it will be *used* more than it's read. This means, most of the time you won't see the "noise" of the comment, but perhaps be glad you can lookup a description from within your IDE instead of guessing the connection to the `iterator` method -or opening the source file to read it. OTOH adding noise to a method name adds noise to all the call sites, so I think those levels of "noise" are hardly comparable.

>  Even repeating method description in @return tag seems pointless to me.

I agree on that, and that's why we often don't include a return tag in newer code, instead using a description that starts with "Returns", eg:

https://github.com/pmd/pmd/blob/1a5486dc3084447ca567b041f6ece32ef7b927ae/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTClassOrInterfaceDeclaration.java#L129-L133

Comment (pzygielo) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: :+1: for all arguments but one

> Possible uses of a javadoc comment here:

In original proposal: https://github.com/pmd/pmd/blob/fb6bb8fee51e14b0ca1ebc00f8903551d81a2ac2/pmd-core/src/main/java/net/sourceforge/pmd/Report.java#L521-L528

Which - assuming meaning of `getNumViolations` is well-established (supposedly it's only me, but I will know since now as well - additional points for _if the int return type is not explicit enough_) is unnecessary. Like for referenced `isEmpty` method.

> Clarify the occasional confusion you describe with get violations of number(s), by simply saying this is the "number of violations"

Good point. Modified javadoc then, not repeating content, would be the best.

Comment (pdhung3012) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: Thank you. I updated the javadoc.

Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: > Which - assuming meaning of getNumViolations is well-established (supposedly it's only me, but I will know since now as well - additional points for if the int return type is not explicit enough) is unnecessary. Like for referenced isEmpty method.

I don't really understand this paragraph. What are you saying is unnecessary? 

I think eg a comment like the following is simple enough and doesn't repeat content:
```java
/**
 * Returns the number of reported violations. 
 * This is the number of elements yielded by the {@link #iterator()} method. 
 */
```
The clarification for the confusion is kind of implicit, by just describing what the method returns in plain English, rather than making a separate sentence.

Btw the `isEmpty` method would IMO also need a javadoc to clarify what it does. It's not just the plain `isEmpty` you'd find on a `List<RuleViolation>`, since it also checks the number of errors. Just relying on the name here, you could write
```java
if (!report.isEmpty()) {
 RuleViolation first = report.iterator().next();
}
```
and get a `NoSuchElementException` if there are only errors.

Comment (pzygielo) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: > I don't really understand this paragraph. What are you saying is unnecessary?

Then I meant javadoc. (My fault of writing sentence too complex for myself.)

Your comment for that new method seems very fine.

----
> <sup>Btw the isEmpty method would IMO also need a javadoc to clarify what it does.</sup>

<sup>So is the name correct? But let's not discuss it here and now, as it's not its review.</sup>

Comment (pzygielo) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: > I think `getNumX` is such a well-known convention that there's hardly any ambiguity. Eg in JDK 10, there are ~322~ 179 method declarations with this pattern, versus 55 for `getNumberOfX`. It's also used more in our own sources (19 vs 6).

So true inconsistency/free-style in JDK alone...
[ColorSpace.getNumComponents](https://docs.oracle.com/en/java/javase/12/docs/api/java.desktop/java/awt/color/ColorSpace.html#getNumComponents()) vs. [MouseInfo.getNumberOfButtons](https://docs.oracle.com/en/java/javase/12/docs/api/java.desktop/java/awt/MouseInfo.html#getNumberOfButtons()). Thanks for pointing to Sun. I'm glad I've never used AWT.
But I won't sleep well. :disappointed:

Comment (pdhung3012) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: Is there any changes needed from my side on this pull request? Thanks

Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/Report.java]: @pdhung3012 Well after reviewing the Report API in #1962, and assuming we follow through on it, this PR won't be needed, as `size()` would be deprecated with no replacement anyway. Until we decide on it, this PR can be considered on hold.

**Comment by adangel** [comment]
Replaced by #2680
